pkg/util/watch/until.go

```diff
@@ -0,0 +1,403 @@
+// are converted into ListOptions and merged with any specified ListOptions.
+type flexWatcher struct {
+	// Context to pass to client list and watch methods
+	Context context.Context
+
+	// Client to list and watch with
+	Client client.WithWatch
+
+	// Key is the name & namespace to watch (both optional)
+	Key client.ObjectKey
+
+	// Labels to filter with (optional)
+	Labels map[string]string
+
+	// ObjList is the collection type to list & watch
+	ObjList client.ObjectList
+}
+
+// Watch satisfies the cache.Lister interface
+func (fw *flexWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	objList, err := kinds.ObjectAsClientObjectList(fw.ObjList.DeepCopyObject())
+	if err != nil {
+		return nil, err
+	}
+	cOpts, err := ConvertListOptions(&options)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert metav1.ListOptions to client.ListOptions")
+	}
+	cOpts, err = MergeListOptions(cOpts, fw.listOptions())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to merge list options")
+	}
+	optsList := UnrollListOptions(cOpts)
+	klog.V(5).Infof("Listing %T %s: %v", objList, objList.GetObjectKind().GroupVersionKind().Kind, optsList)
+	err = fw.Client.List(fw.Context, objList, optsList...)
+	return objList, err
+}
+
+// Watch satisfies the cache.Watcher interface
+func (fw *flexWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	objList, err := kinds.ObjectAsClientObjectList(fw.ObjList.DeepCopyObject())
+	if err != nil {
+		return nil, err
+	}
+	cOpts, err := ConvertListOptions(&options)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to convert metav1.ListOptions to client.ListOptions")
+	}
+	cOpts, err = MergeListOptions(cOpts, fw.listOptions())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to merge list options")
+	}
+	optsList := UnrollListOptions(cOpts)
+	klog.V(5).Infof("Watching %T %s: %v", objList, objList.GetObjectKind().GroupVersionKind().Kind, optsList)
+	return fw.Client.Watch(fw.Context, objList, optsList...)
+}
+
+// listOptions builds a set of List filters for the flexWatcher's Key and Labels.
+func (fw *flexWatcher) listOptions() *client.ListOptions {
+	opts := client.ListOptions{}
+	if len(fw.Key.Namespace) > 0 {
+		opts.Namespace = fw.Key.Namespace
+	}
+	if len(fw.Key.Name) > 0 {
+		opts.FieldSelector = fields.OneTermEqualSelector(metav1.ObjectNameField, fw.Key.Name)
+	}
+	if len(fw.Labels) > 0 {
+		opts.LabelSelector = client.MatchingLabelsSelector{
+			Selector: labels.SelectorFromSet(fw.Labels),
+		}
+	}
+	return &opts
+}
+
+// MergeListOptions merges two sets of ListOptions.
+// - For Namespace, Limit, and Continue, at most one may be specified, unless they're the same
+// - For FieldSelector and LabelSelector, the result requires both filters to match (AND)
+// - For Raw, at most one may be specified
+func MergeListOptions(a, b *client.ListOptions) (*client.ListOptions, error) {
```

Comment:
These util functions could use some unit tests

Reply:
yes mother

Reply:
![image](https://media.tenor.com/Hgwx2qpY-K8AAAAC/mom-i-am-your-mother.gif)

Reply:
my room is clean now, mother.